File: agent/internal/protocol/protocol.go
// Package protocol defines the wire format between the Dockplane agent and the
// control server.
//
// The definitions mirror the server's. Both sides validate independently: the
// agent does not assume a message is safe because the server sent it, and the
// server does not assume a reply is well formed because an authenticated agent
// produced it.
package protocol

import (
	"encoding/json"
	"errors"
	"fmt"
	"time"
)

// Version is the protocol this agent speaks. It is authoritative for
// compatibility; the agent build version is informational.
const Version = 1

// MaxMessageBytes bounds a single message. A peer that never sends a delimiter
// must not be able to grow the receive buffer without limit.
const MaxMessageBytes = 1 << 20

// Message types sent by the agent.
const (
	TypeHello         = "hello"
	TypeHeartbeat     = "heartbeat"
	TypeRenew         = "certificate.renew"
	TypeResponse      = "response"
	TypeStreamStarted = "stream_started"
	TypeStreamChunk   = "stream_chunk"
	TypeStreamEnd     = "stream_end"
)

// Message types sent by the server.
const (
	TypeHelloAck           = "hello_ack"
	TypeHeartbeatAck       = "heartbeat_ack"
	TypeCertificateRenewed = "certificate.renewed"
	TypeRequest            = "request"
	TypeStreamCancel       = "stream_cancel"
	TypeError              = "error"
)

// Why a stream ended. The agent reports one of these and nothing else.
const (
	StreamCompleted = "completed"
	StreamCancelled = "cancelled"
	StreamFailed    = "failed"
	StreamExpired   = "expired"
)

// Hello opens a connection. It carries no identity: the server derives that
// from the client certificate that completed the TLS handshake.
type Hello struct {
	Type            string   `json:"type"`
	ProtocolVersion int      `json:"protocolVersion"`
	AgentVersion    string   `json:"agentVersion,omitempty"`
	Capabilities    []string `json:"capabilities,omitempty"`
}

// Heartbeat reports liveness.
type Heartbeat struct {
	Type            string `json:"type"`
	ProtocolVersion int    `json:"protocolVersion"`
}

// RenewCertificate asks for a replacement certificate. It carries no token: the
// current certificate authenticated the connection it arrives on.
type RenewCertificate struct {
	Type            string `json:"type"`
	ProtocolVersion int    `json:"protocolVersion"`
	CSR             string `json:"csr"`
}

// Response answers a capability request.
type Response struct {
	Type            string         `json:"type"`
	ProtocolVersion int            `json:"protocolVersion"`
	ID              string         `json:"id"`
	Capability      string         `json:"capability"`
	Status          string         `json:"status"`
	Payload         any            `json:"payload,omitempty"`
	Error           *ResponseError `json:"error,omitempty"`
}

// ResponseError reports why a capability failed.
type ResponseError struct {
	Code    string `json:"code"`
	Message string `json:"message"`
}

/*
StreamStarted acknowledges that a streaming capability is running.

It carries the identifier the server assigned rather than one the agent chose,
so the server can bind what follows to the stream it opened.
*/
type StreamStarted struct {
	Type            string `json:"type"`
	ProtocolVersion int    `json:"protocolVersion"`
	ID              string `json:"id"`
	Capability      string `json:"capability"`
	StreamID        string `json:"streamId"`
}

// StreamChunk carries one delivery of a running stream.
type StreamChunk struct {
	Type            string `json:"type"`
	ProtocolVersion int    `json:"protocolVersion"`
	ID              string `json:"id"`
	StreamID        string `json:"streamId"`
	// Seq numbers the chunks of one stream from zero, so a gap is visible.
	Seq     int `json:"seq"`
	Payload any `json:"payload"`
	// Dropped counts what the agent discarded because the consumer was behind.
	Dropped int `json:"dropped,omitempty"`
}

// StreamEnd closes a stream. Nothing more may be sent for that identifier.
type StreamEnd struct {
	Type            string         `json:"type"`
	ProtocolVersion int            `json:"protocolVersion"`
	ID              string         `json:"id"`
	StreamID        string         `json:"streamId"`
	Reason          string         `json:"reason"`
	Error           *ResponseError `json:"error,omitempty"`
}

// ServerMessage is any message the server may send. Fields are read only after
// the type has been established.
type ServerMessage struct {
	Type                     string          `json:"type"`
	ProtocolVersion          int             `json:"protocolVersion"`
	AgentID                  string          `json:"agentId"`
	HeartbeatIntervalSeconds int             `json:"heartbeatIntervalSeconds"`
	CertificateNotAfter      string          `json:"certificateNotAfter"`
	RenewAfter               string          `json:"renewAfter"`
	Certificate              string          `json:"certificate"`
	ID                       string          `json:"id"`
	StreamID                 string          `json:"streamId"`
	Capability               string          `json:"capability"`
	IssuedAt                 string          `json:"issuedAt"`
	ExpiresAt                string          `json:"expiresAt"`
	Payload                  json.RawMessage `json:"payload"`
	Code                     string          `json:"code"`
	Message                  string          `json:"message"`
}

// Errors returned when a server message cannot be trusted.
var (
	ErrUnsupportedVersion = errors.New("unsupported protocol version")
	ErrMalformed          = errors.New("malformed message")
	ErrExpired            = errors.New("request expired")
)

// ParseServerMessage decodes one message and checks what every message must
// carry. Type-specific validation belongs to the caller that acts on it.
func ParseServerMessage(line []byte) (*ServerMessage, error) {
	if len(line) > MaxMessageBytes {
		return nil, fmt.Errorf("%w: %d bytes", ErrMalformed, len(line))
	}

	var message ServerMessage

	if err := json.Unmarshal(line, &message); err != nil {
		return nil, fmt.Errorf("%w: %v", ErrMalformed, err)
	}

	if message.Type == "" {
		return nil, fmt.Errorf("%w: no type", ErrMalformed)
	}

	if message.ProtocolVersion != Version {
		return nil, fmt.Errorf("%w: %d", ErrUnsupportedVersion, message.ProtocolVersion)
	}

	return &message, nil
}

// ValidateRequest checks the envelope of a capability request.
//
// The expiry is enforced by the agent as well as by the server. A request that
// waited in a queue, or was replayed, must not run late against a host whose
// state has moved on.
func ValidateRequest(message *ServerMessage, now time.Time) error {
	if message.ID == "" || message.Capability == "" {
		return fmt.Errorf("%w: request without an identifier or capability", ErrMalformed)
	}

	issuedAt, err := time.Parse(time.RFC3339, message.IssuedAt)
	if err != nil {
		return fmt.Errorf("%w: issuedAt", ErrMalformed)
	}

	expiresAt, err := time.Parse(time.RFC3339, message.ExpiresAt)
	if err != nil {
		return fmt.Errorf("%w: expiresAt", ErrMalformed)
	}

	if !expiresAt.After(issuedAt) {
		return fmt.Errorf("%w: expiry is not after issue", ErrMalformed)
	}

	if now.After(expiresAt) {
		return fmt.Errorf("%w: expired at %s", ErrExpired, expiresAt.Format(time.RFC3339))
	}

	return nil
}

// NewSuccess builds a successful reply. The capability is echoed so the server
// can refuse a reply that answers a different question than it asked.
func NewSuccess(id, capability string, payload any) Response {
	return Response{
		Type:            TypeResponse,
		ProtocolVersion: Version,
		ID:              id,
		Capability:      capability,
		Status:          "success",
		Payload:         payload,
	}
}

// NewStreamStarted acknowledges a stream the server asked for.
func NewStreamStarted(id, capability, streamID string) StreamStarted {
	return StreamStarted{
		Type:            TypeStreamStarted,
		ProtocolVersion: Version,
		ID:              id,
		Capability:      capability,
		StreamID:        streamID,
	}
}

// NewStreamChunk builds one delivery.
func NewStreamChunk(id, streamID string, seq int, payload any, dropped int) StreamChunk {
	return StreamChunk{
		Type:            TypeStreamChunk,
		ProtocolVersion: Version,
		ID:              id,
		StreamID:        streamID,
		Seq:             seq,
		Payload:         payload,
		Dropped:         dropped,
	}
}

// NewStreamEnd closes a stream, with a code when it ended badly.
func NewStreamEnd(id, streamID, reason string, failure *ResponseError) StreamEnd {
	return StreamEnd{
		Type:            TypeStreamEnd,
		ProtocolVersion: Version,
		ID:              id,
		StreamID:        streamID,
		Reason:          reason,
		Error:           failure,
	}
}

// NewFailure builds a failed reply.
func NewFailure(id, capability, code, message string) Response {
	return Response{
		Type:            TypeResponse,
		ProtocolVersion: Version,
		ID:              id,
		Capability:      capability,
		Status:          "error",
		Error:           &ResponseError{Code: code, Message: message},
	}
}

File: api/src/agents/protocol.ts
/**
 * Agent gateway protocol.
 *
 * Two kinds of traffic share one connection: liveness, which the agent starts,
 * and capability requests, which the server starts. Both are versioned, and
 * every capability request carries its own identifier and expiry, so a reply
 * can be correlated and a stale request can be refused rather than run late.
 */

import { Capability, isCapability } from './capabilities';

export const PROTOCOL_VERSION = 1;
export const MINIMUM_PROTOCOL_VERSION = 1;

export function isSupportedProtocolVersion(version: unknown): version is number {
  return (
    typeof version === 'number' &&
    Number.isInteger(version) &&
    version >= MINIMUM_PROTOCOL_VERSION &&
    version <= PROTOCOL_VERSION
  );
}

export interface HelloMessage {
  readonly type: 'hello';
  readonly protocolVersion: number;
  readonly agentVersion?: string;
  /** Advertised by the agent. None are honoured in this milestone. */
  readonly capabilities?: readonly string[];
}

export interface HeartbeatMessage {
  readonly type: 'heartbeat';
  readonly protocolVersion: number;
}

export interface CertificateRenewalMessage {
  readonly type: 'certificate.renew';
  readonly protocolVersion: number;
  readonly csr: string;
}

/**
 * A capability result.
 *
 * The capability is echoed so the server can refuse a reply that answers a
 * different question than the one it asked, even when the identifier matches.
 */
export interface CapabilityResponseMessage {
  readonly type: 'response';
  readonly protocolVersion: number;
  readonly id: string;
  readonly capability: Capability;
  readonly status: 'success' | 'error';
  readonly payload?: unknown;
  readonly error?: { readonly code: string; readonly message: string };
}

/*
 * A streaming capability reports its progress in three parts.
 *
 * The stream identifier is the server's, sent with the request and echoed
 * back, so everything that arrives can be bound to the stream it opened rather
 * than to one the agent named.
 */
export interface StreamStartedMessage {
  readonly type: 'stream_started';
  readonly protocolVersion: number;
  readonly id: string;
  readonly capability: Capability;
  readonly streamId: string;
}

export interface StreamChunkMessage {
  readonly type: 'stream_chunk';
  readonly protocolVersion: number;
  readonly id: string;
  readonly streamId: string;
  /** Numbered from zero, so a gap in delivery is visible. */
  readonly seq: number;
  readonly payload: unknown;
  /** What the agent discarded because the consumer was behind. */
  readonly dropped: number;
}

export interface StreamEndMessage {
  readonly type: 'stream_end';
  readonly protocolVersion: number;
  readonly id: string;
  readonly streamId: string;
  readonly reason: StreamEndReason;
  readonly error?: { readonly code: string; readonly message: string };
}

export const STREAM_END_REASONS = ['completed', 'cancelled', 'failed', 'expired'] as const;

export type StreamEndReason = (typeof STREAM_END_REASONS)[number];

const END_REASONS = new Set<string>(STREAM_END_REASONS);

export type ClientMessage =
  | HelloMessage
  | HeartbeatMessage
  | CertificateRenewalMessage
  | CapabilityResponseMessage
  | StreamStartedMessage
  | StreamChunkMessage
  | StreamEndMessage;

/** A capability the server asks an agent to perform. */
export interface CapabilityRequestMessage {
  readonly type: 'request';
  readonly protocolVersion: number;
  readonly id: string;
  readonly capability: Capability;
  readonly issuedAt: string;
  readonly expiresAt: string;
  readonly payload: Record<string, unknown>;
  /** Present only for a streaming capability, and assigned by the server. */
  readonly streamId?: string;
}

/** Ends a stream the server no longer wants. */
export interface StreamCancelMessage {
  readonly type: 'stream_cancel';
  readonly protocolVersion: number;
  readonly id: string;
  readonly streamId: string;
}

export interface HelloAck {
  readonly type: 'hello_ack';
  readonly protocolVersion: number;
  readonly agentId: string;
  readonly heartbeatIntervalSeconds: number;
  readonly certificateNotAfter: string;
  /** When the agent should start renewing. The server owns this policy. */
  readonly renewAfter: string;
}

export interface HeartbeatAck {
  readonly type: 'heartbeat_ack';
  readonly protocolVersion: number;
}

export interface CertificateRenewed {
  readonly type: 'certificate.renewed';
  readonly protocolVersion: number;
  readonly certificate: string;
  readonly certificateNotAfter: string;
  readonly renewAfter: string;
}

export interface ProtocolError {
  readonly type: 'error';
  readonly protocolVersion: number;
  readonly code: string;
  readonly message: string;
}

export type ServerMessage =
  | HelloAck
  | HeartbeatAck
  | CertificateRenewed
  | ProtocolError
  | CapabilityRequestMessage
  | StreamCancelMessage;

export const HEARTBEAT_INTERVAL_SECONDS = 30;

/**
 * Parses one message.
 *
 * Anything that is not a known object with a known type is rejected outright;
 * the gateway never infers intent from a partially understood payload.
 */
export function parseClientMessage(raw: string): ClientMessage | undefined {
  let value: unknown;

  try {
    value = JSON.parse(raw);
  } catch {
    return undefined;
  }

  if (typeof value !== 'object' || value === null) {
    return undefined;
  }

  const message = value as Record<string, unknown>;

  if (typeof message.type !== 'string' || typeof message.protocolVersion !== 'number') {
    return undefined;
  }

  switch (message.type) {
    case 'hello':
      return {
        type: 'hello',
        protocolVersion: message.protocolVersion,
        agentVersion: typeof message.agentVersion === 'string' ? message.agentVersion : undefined,
        capabilities: Array.isArray(message.capabilities)
          ? message.capabilities.filter((entry): entry is string => typeof entry === 'string')
          : [],
      };

    case 'heartbeat':
      return { type: 'heartbeat', protocolVersion: message.protocolVersion };

    case 'certificate.renew':
      return typeof message.csr === 'string'
        ? { type: 'certificate.renew', protocolVersion: message.protocolVersion, csr: message.csr }
        : undefined;

    case 'response':
      return parseCapabilityResponse(message);

    case 'stream_started':
      return typeof message.id === 'string' &&
        typeof message.streamId === 'string' &&
        isCapability(message.capability)
        ? {
            type: 'stream_started',
            protocolVersion: message.protocolVersion as number,
            id: message.id,
            capability: message.capability,
            streamId: message.streamId,
          }
        : undefined;

    case 'stream_chunk':
      return parseStreamChunk(message);

    case 'stream_end':
      return typeof message.id === 'string' &&
        typeof message.streamId === 'string' &&
        typeof message.reason === 'string' &&
        END_REASONS.has(message.reason)
        ? {
            type: 'stream_end',
            protocolVersion: message.protocolVersion as number,
            id: message.id,
            streamId: message.streamId,
            reason: message.reason as StreamEndReason,
            error: parseError(message.error),
          }
        : undefined;

    default:
      return undefined;
  }
}

function parseStreamChunk(message: Record<string, unknown>): StreamChunkMessage | undefined {
  if (typeof message.id !== 'string' || typeof message.streamId !== 'string') {
    return undefined;
  }

  if (typeof message.seq !== 'number' || !Number.isInteger(message.seq) || message.seq < 0) {
    return undefined;
  }

  const dropped = typeof message.dropped === 'number' ? message.dropped : 0;

  return {
    type: 'stream_chunk',
    protocolVersion: message.protocolVersion as number,
    id: message.id,
    streamId: message.streamId,
    seq: message.seq,
    payload: message.payload,
    dropped: Number.isInteger(dropped) && dropped > 0 ? dropped : 0,
  };
}

function parseError(value: unknown): { code: string; message: string } | undefined {
  if (typeof value !== 'object' || value === null) {
    return undefined;
  }

  const error = value as Record<string, unknown>;

  return {
    code: typeof error.code === 'string' ? error.code : 'AGENT_CAPABILITY_FAILED',
    message: typeof error.message === 'string' ? error.message : 'The capability failed.',
  };
}

function parseCapabilityResponse(
  message: Record<string, unknown>,
): CapabilityResponseMessage | undefined {
  if (typeof message.id !== 'string' || !isCapability(message.capability)) {
    return undefined;
  }

  if (message.status !== 'success' && message.status !== 'error') {
    return undefined;
  }

  return {
    type: 'response',
    protocolVersion: message.protocolVersion as number,
    id: message.id,
    capability: message.capability,
    status: message.status,
    payload: message.payload,
    error: parseError(message.error),
  };
}
